rust/nix-expr/src/eval_state.rs

```diff
@@ -34,24 +36,42 @@ pub fn init() -> Result<()> {
     }
 }
 
+pub struct RealisedString {
+    pub s: String,
+    pub paths: Vec<StorePath>,
+}
+
 pub struct EvalState {
     eval_state: NonNull<raw::EvalState>,
     store: Store,
     context: Context,
 }
 impl EvalState {
-    pub fn new(store: Store) -> Result<Self> {
+    pub fn new<'a>(store: Store, lookup_path: impl IntoIterator<Item = &'a str>) -> Result<Self> {
         let context = Context::new();
 
+        // this intermediate value must be here and must not be moved
+        // because it owns the data the `*const c_char` pointers point to.
+        let lookup_path: Vec<CString> = lookup_path
+            .into_iter()
+            .map(|path| {
+                CString::new(path).with_context(|| {
+                    format!("EvalState::new: lookup_path `{path}` contains null byte")
+                })
+            })
+            .collect::<Result<_>>()?;
```

Comment:
i do not see how [`Pin`](https://doc.rust-lang.org/std/pin/struct.Pin.html) would help here. see [this section in the `std::pin` documentation](https://doc.rust-lang.org/std/pin/index.html#pinning-as-a-library-contract).

Reply:
Thanks for the response. I had not read through the documentation on `Pin` enough to understand what's going on. Having read through it now, I can say adding `Pin` won't help in automatically checking anything. Anything we do with the value would need to be made `unsafe {}` and anyways be checked that they're not moving the value.

WIth that in mind, here's my pros/cons list for including `Pin`.

## Pros

- Clearly communicate what the constraints on the value are (like this comment is trying to do)
- If something gets refactored, it may be harder to accidentally include a move.
- It would mean marking places where the "this must not be moved" invariant must be manually upheld.

## Cons

- Makes things more complicated
  - You Ain't Gonna Need It
  - This code is just not complicated enough to warrant this kind of mechanism
- It would show up as `Vec<Pin<*const c_char>>` below, but the `Vec<CString>` value above could still be moved.
  - Though in practice it's being shadowed so there's no way to refer to it.
  
 I'm leaning slightly towards "leave it as it is", but I don't feel strongly.

Reply:
Another thing you could do that helps communicate "these code pieces are part of a unit and must not be moved away from each other" is to put them in an assignment block.

```rust
        let lookup_path: *mut *const c_char = {
            // this intermediate value must be here and must not be moved
            // because it owns the data the `*const c_char` pointers point to.
            let lookup_path: Vec<CString> = lookup_path
                .into_iter()
                .map(|path| {
                    CString::new(path).with_context(|| {
                        format!("EvalState::new: lookup_path `{path}` contains null byte")
                    })
                })
                .collect::<Result<_>>()?;

            // this intermediate value owns the data the `*mut *const c_char` pointer points to.
            let mut lookup_path: Vec<*const c_char> = lookup_path
                .iter()
                .map(|s| s.as_ptr())
                .chain(std::iter::once(null())) // signal the end of the array
                .collect();

            lookup_path.as_mut_ptr()
        };

```

Reply:
this causes a segmentation fault because the intermediate values owning the data are dropped before the c api call! is this not incredibly subtle? i got burned by this a lot.

Reply:
this demonstrates where i could improve my comment too.
> this intermediate value must be here and must neither be moved nor dropped before the api call to `state_create` because it owns the data the `*const c_char` pointers point to.

Reply:
> this causes a segmentation fault 

Ah damn. I was afraid of this, but thought it would be a lifetimes issue so when the LSP didn't complain I thought it was fine.

At this point I don't think there's a better way, but this section of code is rather delicate. I suppose that's what the comment is for.